File: DataStructure/DoubleLinkedList/DoubleLinkedList.py
# DoubleLinkedList implementation in Python

# This module implements a simple DoubleLikedList with basic operations.
class Node:
    # Declare a Node consisting of data and two pointer prevent, next node
    def __init__(self, data) -> None: # type: ignore
        self.data = data # type: ignore
        self.prev = None
        self.next = None

# This class implements a DoubleLinkedList with basic operations
class DoubleLinkedList:

    # Declare the head of the DoubleLinkedList
    def __init__(self) -> None:
        self.head = None # Initialize the head of the DoubleLinkedList
        self._size = 0 # type: ignore
    
    # Method to add a new node at the end of the DoubleLinkedList
    def append(self, data) -> None: # type: ignore
        new_node = Node(data) # type: ignore # Create a new node with given data
        if not self.head: #If the DoubleLinkedList is empty
            self.head = new_node
        else:
            current = self.head # Start from the head
            while current.next:
                current = current.next
            current.next = new_node # type: ignore
            new_node.prev = current # type: ignore
        self._size += 1 # type: ignore

    # Method to delete a node with the specified data from the DoubleLinkedList
    def delete(self, data) -> None: # type: ignore
        current = self.head
        while current: # Traverse through the list
            if current.data == data: # type: ignore # If the current node's data matches
                if current.prev: # middle or last node
                    current.prev.next = current.next
                else: # first node
                    self.head = current.next
                if current.next: 
                    current.next.prev = current.prev
                self._size -= 1 # type: ignore
                return
            current = current.next # Move to the next node
    
    # Method to find with the specified data in the DoubleLinkedList
    def find(self, data): # type: ignore
        current = self.head
        while current: # Traverse through the list
            if current.data==data: # type: ignore # If the current node's data matches
                return current # Return the current node
            current = current.next
        return None # If not found, return None
    
    # Method to get the size of the DoubleLindedList
    def size(self) -> int:
        return self._size  # type: ignore
    
    # Method to print the DoubleLinkedList
    def print_list(self):
        current = self.head
        tail = None # type: ignore

        print("Forward : ", end="")
        while current: # type: ignore
            print(current.data, end=" -> ") # type: ignore
            tail = current # type: ignore
            current = current.next
        print("Null")

        print("Backward : ", end="")
        while tail:
            print(tail.data, end=" <- ") # type: ignore
            tail = tail.prev
        print("head")

# Example usage
if __name__ == "__main__":
    dll = DoubleLinkedList() # Create a new LinkedList instance
    dll.append(1) # type: ignore # Append data 1
    dll.append(2) # type: ignore # Append data 2
    dll.append(3) # type: ignore # Append data 3
    dll.append(4) # type: ignore # Append data 4
    dll.append(5) # type: ignore # Append data 5

    dll.print_list() # Print the linked list
    print(dll.find(3)) # type: ignore # Find a node with data 3
    dll.delete(3) # type: ignore # Delete the node with data 3
    dll.print_list() # Print the linked list after deletion 

    found_node = dll.find(3) # type: ignore # Find a node with data 3
    if found_node:
        print(f"Found node with data: {found_node.data}") # type: ignore
    else:
        print("Node not found")

File: DataStructure/DoubleLinkedList/DoubleLinkedList.js
console.log("Executing from:", process.cwd());
// DoubleLinkedList implementation in JavaScript

// This module implements a simple DoubleLinkedList with basic operations
class Node{

    // Declare a Node consisting of data and two pointer prevent, next node
    constructor(data) {
        this.data = data;
        this.prev = null;
        this.next = null;
    }
}

// This class implements a DoubleLinkedList with basic operations
class DoubleLinkedList{

    // Declare the head of the DoubleLinkedList
    constructor() {
        this.head = null;
        this._size = 0;
    }

    // Method to add a new node at the end of the DoubleLinkedList
    append(data) {
        const new_node = new Node(data)
        if(this.head === null) { this.head = new_node }
        else {
            let current = this.head
            while(current.next != null) {
                current = current.next;
            }
            current.next = new_node;
            new_node.prev = current;
        }
        this._size++;
    }

    // Method to delete a note with the specified data from in DoubleLinkedList
    delete(data) {
        if(!this.head) return; // head is empty
        
        let current = this.head;

        while(current) { // Traverse through the list
            if(current.data === data) { // If the current node's data matches
                if(current === this.head) { // Data match first node
                    this.head = current.next;
                    if(this.head) { this.head.prev = null; }
                } else {// Data match other node
                    if(current.next) { current.next.prev = current.prev; }
                    current.prev.next = current.next;
                }
                this._size--;
            }
            current = current.next;
        }
    }

    // Method to find with the specified data in the DoubleLinkedList.
    find(data) {
        if(!this.head) return;

        let current = this.head;

        while(current) {
            if(current.data === data) {return current; }
            else { current=current.next; }
        }

        return null;
    }

    //Method to get the size of the DoubleLinkedList
    size() { return this._size; }

    //Method to print the DoubleLinkedList
    print() {
        if(!this.head) return console.log("Empty!");
        
        let current = this.head;
        let tail = null;
        let list_str = "";
        
        list_str = "Forward : head";
        while(current) {
            list_str += (current.data + " -> ");
            tail = current;
            current = current.next;
        }
        list_str += "null"
        console.log(list_str);
        list_str = "";

        list_str = "Backword : ";
        while(tail) {
            list_str += (tail.data + " <- ");
            tail = tail.prev;
        }
        list_str += "head "
        console.log(list_str);
    }
}

// Example usage
const doublelist = new DoubleLinkedList();
doublelist.append(10);
doublelist.append(20);
doublelist.append(30);
doublelist.append(40);
doublelist.append(50);
doublelist.print(); // Output: 10 -> 20 -> 30 -> 40 -> 50 -> null
doublelist.delete(30);
doublelist.print(); // Output: 10 -> 20 -> 40 -> 50 -> null
console.log(doublelist.find(20)); // Output: Node { data: 20, next: null }
console.log(doublelist.size()); // Output: 4